### PR TITLE
Add AWS region support

### PR DESCRIPTION
**Pull request by miquella**
This PR adds the ability to specify a region in the vault itself or on the CLI via the `--region …` flag. Thankfully, the AWS SDK handles most of the heavy lifting as long as the region is specified (including selecting the correct partition and endpoints).

Unfortunately, the standard partition defaults to using the global STS endpoint rather than respecting the region specification, so a custom resolver had to be implemented to select region-local endpoints. However, the global STS endpoint is still used if no region is specified.

To help smooth over the user experience, the session cache has been upgraded to be able to cache sessions from multiple different variations of a vault. (i.e. different regions or similar)

### CODE REVIEW

**Comment by miquella** [comment]
This should cover #142 as well.

File: command.go [hunk 1]
```diff
@@ -267,6 +267,7 @@ func parseEnvArgs(args []string) (Command, error) {
 	flag.String("assume", "", "Role to assume")
 	flag.Bool("no-session", false, "Disable use of temporary credentials")
 	flag.Bool("refresh", false, "Start a new session with new temporary credentials and a refreshed expiration")
+	flag.String("region", os.Getenv("AWS_DEFAULT_REGION"), "The AWS region to use to generate STS credentials")
```
Comment (tpickett66) [inline command.go]: Isn't `AWS_REGION` the preferred variable here with this being the fallback?

File: command.go [hunk 2]
```diff
@@ -460,6 +462,7 @@ func parseShellArgs(args []string) (Command, error) {
...
```
Comment (tpickett66) [inline command.go]: Same comment here about `AWS_REGION` vs `AWS_DEFAULT_REGION`

File: lib/aws_credentials.go [hunk 3]
```diff
@@ -106,19 +113,27 @@ func (c *AWSCredentials) AssumeRole(arn string, duration time.Duration) (*AWSCre
...
 	config := &aws.Config{}
-	if c != nil && c.ID != "" {
-		config.Credentials = credentials.NewStaticCredentials(
-			c.ID,
-			c.Secret,
-			c.Token,
-		)
+
+	if c != nil {
+		if c.Region != nil {
```
Comment (tpickett66) [inline lib/aws_credentials.go]: You could push this nil check down a level if you made your custom resolver return the global endpoint when no region is set. I don't know if that gains us much but it's an option.

File: lib/session.go [hunk 4]
```diff
@@ -252,6 +276,10 @@ func (s *Session) Variables() *Variables {
 				"AWS_SECURITY_TOKEN",
 			)
 		}
+
+		if s.AWSCreds.Region != nil && *s.AWSCreds.Region != "" {
+			vars.Set["AWS_DEFAULT_REGION"] = *s.AWSCreds.Region
```
Comment (tpickett66) [inline lib/session.go]: Same question here about `AWS_REGION` vs `AWS_DEFAULT_REGION`

File: menu/aws.go [hunk 5]
```diff
@@ -164,6 +195,19 @@ func (m *AWSMenu) Handler() error {
 	}
 }
 
+func (m *AWSMenu) readRegion() (string, error) {
+	region, err := interaction.ReadValue("Region: ")
+	if err != nil {
+		return "", err
+	}
+
+	if !sdkRegions()[region] {
```
Comment (tpickett66) [inline menu/aws.go]: I think this should be an error rather than just a warning to the user.

File: command.go (hunk 1 above)
Comment (miquella) [reply command.go]: After talking with some AWS folks, it appears that there isn't a consensus on which variable should be used (some tools completely ignore one variable or another), but I think I agree with your premise and will update the code to prefer `AWS_REGION` and fall back to `AWS_DEFAULT_REGION`.

File: command.go (hunk 2 above)
Comment (miquella) [reply command.go]: Adopting as noted in above comment.

File: lib/aws_credentials.go (hunk 3 above)
Comment (miquella) [reply lib/aws_credentials.go]: Good call. This was originally written this way before I realized a custom resolver was necessary.

File: lib/session.go (hunk 4 above)
Comment (miquella) [reply lib/session.go]: As mentioned in my other comment, AWS tools don't appear to have standardized on one variable or the other. I'll update this to export both.

File: menu/aws.go (hunk 5 above)
Comment (miquella) [reply menu/aws.go]: This is actually intentional to allow Vaulted to work with a new region before the SDK is updated to support it.

But to help ensure that the user doesn't miss the message, line [245](https://github.com/miquella/vaulted/pull/149/files/abbe71178ff751a5e252e2cb62c963659137d9bf#diff-65c99fdf6b74bed7314728ebece8a0cdR245) ensures that each time the vault contents are rendered, a message indicating `(unrecognized region)` is displayed.

File: session_generation.go [hunk 6]
```diff
@@ -100,3 +112,28 @@ func getVaultSession(store vaulted.Store, options *SessionOptions) (*vaulted.Ses
 	// Assume the session's role
 	return session.AssumeSessionRole()
 }
+
+func updateVaultFromEnvAndOptions(vault *vaulted.Vault, options *SessionOptions) {
+	// Calculate the region (lowest precedence to highest)
+	region := os.Getenv("AWS_DEFAULT_REGION")
```
Comment (tpickett66) [inline session_generation.go]: I would have put these in the opposite order but I'll allow it.

Comment (miquella) [reply session_generation.go]: I was trying to follow prior art :smile: :

* [Ruby SDK](https://github.com/aws/aws-sdk-ruby/blob/da245ccb2cf9260cee0c6822677d6caf738f7cc3/gems/aws-sdk-core/lib/aws-sdk-core/plugins/regional_endpoint.rb#L50)
* [Go SDK](https://github.com/aws/aws-sdk-go/blob/60a7f0014f86142c03773706dd0cac474570e810/aws/session/env_config.go#L159)

**Comment by miquella** [comment]
Latest patchset corrects an issue with the endpoint resolver in the case that no region is provided.

File: lib/session_cache.go [hunk 7]
```diff
@@ -0,0 +1,120 @@
...
+	//
+	// Any cache loaded that does not match this version is ignored. This
+	// causes all caches written for previous versions to be invalidated.
+	SessionCacheVersion = "3"
+)
+
+var (
+	// ErrVaultSessionNotFound occurs when attempting to locate a vault session
+	// in a SessionCache that isn't present.
+	ErrVaultSessionNotFound = errors.New("vault session not found")
```
[the remark this answers is not in the document]
Comment (miquella) [reply lib/session_cache.go]: Agreed. Fixed.

File: lib/session_cache.go [hunk 8]
```diff
@@ -0,0 +1,120 @@
...
+
+var (
+	// SessionCacheVersion indicates the current version of the cache format.
+	//
+	// Any cache loaded that does not match this version is ignored. This
+	// causes all caches written for previous versions to be invalidated.
+	SessionCacheVersion = "3"
+)
+
+var (
```
[the remark this answers is not in the document]
Comment (miquella) [reply lib/session_cache.go]: We certainly could! The decision to not was intentional as this affects how things are grouped in things such as documentation. These felt like logically separate things, so I separated them into different `var` blocks.

File: doc/vaulted-env.1.md [hunk 9]
```diff
@@ -76,6 +76,13 @@ OPTIONS
 `--refresh`
   Start a new session with new temporary credentials and a refreshed expiration.
 
+`--region` *region*
+  Override the region to be used for AWS. This sets the region used when
+  generating temporary credentials.
+
+  When the region is set via `--region` or via the vault itself, the
+  `AWS_DEFAULT_REGION` variable will be set in spawned environments.
```
[the remark this answers is not in the document]
Comment (miquella) [reply doc/vaulted-env.1.md]: Oh, good catch! I knew I was going to miss something when I went back and added `AWS_REGION` support. It really sucks that there is such disparity in the AWS tooling around things like this (variable names, etc.).

**Pull request merged**